doc/src/explanation/solve_output.rst

```diff
@@ -0,0 +1,120 @@
+.. _solve_output:
+
+==========================
+Understanding Solve Output
+==========================
+
+The output of the `solve` function can seem very unwieldy since it may appear to
+arbitrarily return one of 6 different types of output (in addition to raising
+errors). The reasons for this are historical and are biased toward human
+interaction rather than programmatic use. The type of output will depend on the
+type of equation(s) (and how they are entered) and the number of symbols that
+are provided (and how they are provided).
+
+If consistent output from solve is desired use ``dict=True`` to get a list of
+solutions as dictionaries or ``set=True`` to get a tuple containing the list
+of variables and the set of values for those variables. (These settings only
+affect the output when the input does not contain relational expressions.)
+
+    >>> from sympy import sqrt, exp, solve, Symbol, Eq
+    >>> from sympy.abc import x, y, z, a, b
+    >>> solve(x, x, dict=True)
+    [{x: 0}]
+    >>> solve([x**2 - y, x + y - 6], set=True)
+    ([x, y], {(-3, 9), (2, 4)})
+
+The following is an explanation about the conditions under which various
+forms of output that are obtained by *default*.
+
+1. An empty list (or set) indicates that no solution was found.
+
+    >>> [] == solve(sqrt(x) + 1) == solve(sqrt(x) + 1, dict=True)
+    True
+    >>> solve(sqrt(x) + 1, set=True)
+    ([x], set())
+
+2. A list of values is given when the symbol to solve for was
+unambiguous in context because a) the equation was univariate and
+no symbols were specified or b) a single symbol was specified as
+being of interest.
+
+    >>> solve(x**2 - 4)
+    [-2, 2]
+    >>> solve(x - y - 1, x)
+    [y + 1]
+
+3. A single dictionary with keys being symbols and values being the solutions
+for those symbols is the result when equations are passed as a list and are
+all linear in the symbols given. Such a system is automatically generated
+for a single equation (not passed as a list) if there is an undetermined
+coefficients solution for the symbols specified.
```

Comment:
I think there should be a hyphen: "an undetermined-coefficients solution". (May be grammar nitpicking; I do think it makes the wording easier to understand because "an undetermined coefficients" seems to have singular vs. plural disagreement.)